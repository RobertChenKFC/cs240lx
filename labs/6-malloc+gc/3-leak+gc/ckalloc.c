// implement a simple ckalloc/free that adds ckhdr_t to the 
// allocation.
#include "rpi.h"
#include "ckalloc.h"
#include "kr-malloc.h"

unsigned ck_verbose_p = 0;

static hdr_t *alloc_list, *free_list;

// returns pointer to the first allocated header block.
hdr_t *ck_first_alloc(void) {
    return alloc_list;
}
// returns pointer to the first freed header block.
hdr_t *ck_first_free(void) {
    return free_list;
}

// return header associated with <ptr> if one exists.
hdr_t *ck_ptr_is_alloced(void *ptr) {
    for(hdr_t *h = ck_first_alloc(); h; h = ck_next_hdr(h))
        if(ck_ptr_in_block(h,ptr)) 
            return h;
    return 0;
}

/***********************************************************************
 * implement the rest
 */

// is <ptr> inside <h>'s data block?
unsigned ck_ptr_in_block(hdr_t *h, void *ptr) {
    if(h->state != ALLOCED) {
        panic("should only have allocated blocks: sdtate=%d\n", h->state);
        return 0;
    }
    todo("implement this [simple]\n");
}


// free a block allocated with <ckalloc>
void (ckfree)(void *addr, src_loc_t l) {
    hdr_t *h = (void *)addr;
    h -= 1;

    if(h->state != ALLOCED)
        loc_panic(l, "freeing unallocated memory: state=%d\n", h->state);
    if(ck_verbose_p)
        loc_debug(l, "freeing %p\n", addr);

    assert(ck_ptr_is_alloced(addr));
    h->state = FREED;

    todo("implement the rest\n");
    kr_free(h);
}


// interpose on kr_malloc allocations and
//  1. allocate enough space for a header and fill it in.
//  2. add the allocated block to  the allocated list.
void *(ckalloc)(uint32_t nbytes, src_loc_t l) {
    static unsigned block_id=1;

    hdr_t *h = kr_malloc(nbytes + sizeof *h);

    memset(h, 0, sizeof *h);
    h->nbytes_alloc = nbytes;
    h->state = ALLOCED;
    h->alloc_loc = l;
    h->block_id = block_id++;

    todo("implement the rest\n");
}
