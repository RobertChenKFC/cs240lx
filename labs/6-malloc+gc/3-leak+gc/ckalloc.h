// checking allocator.  used for the gc lab.   main functionality:
// given a pointer <p> to an allocated block <b>, can look up the 
// associated header.
#ifndef __CKALLOC_H__
#define __CKALLOC_H__

// helper code for tracking callsite.
#include "src-loc.h"

// a block can be ALLOCED or FREED
typedef enum {  ALLOCED = 11, FREED } state_t;

// these are used for debugging malloc lab
enum { REDZONE_NBYTES = 128, REDZONE_VAL = 0xfe };

// this header is prepended to each allocated block.
typedef struct ck_hdr {
    // next allocated block or next free block.
    struct ck_hdr *next;

    // how many bytes the user requested to allocate: this can be less
    // than the actual amount we allocatd (can't be less!).
    uint32_t nbytes_alloc;  

    // used for checking: record the state of the block.
    // { ALLOCED, FREED }
    uint32_t state;          

    // unique block id: this lets us compare different runs even if the 
    // allocation location changes.  starts at 0.
    uint32_t block_id;

    // see src-loc.h: holds file, fn, lineno that ckalloc was called
    // at.
    src_loc_t alloc_loc;    

    // used for gc: i didn't cksum these.
    uint32_t refs_start;    // number of pointers to the start of the block.
    uint32_t refs_middle;   // number of pointers to the middle of the block.

    uint16_t mark;          // 0 initialize.
    
    // used in the error checking lab.
    uint8_t rz1[REDZONE_NBYTES];
} hdr_t;

// requires that <p> points to the start of the allocated region.
static inline unsigned ck_blk_id(void *p) {
    hdr_t *h = p;
    return h[-1].block_id;
}

// returns pointer to the first allocated header block.
hdr_t *ck_first_alloc(void);

// returns pointer to the first free'd header block.
hdr_t *ck_first_free(void);

// given header <h>: returns pointer to next hdr
// or 0 if none.
static inline hdr_t *ck_next_hdr(hdr_t *p) {
    return p ? p->next : 0;
}

// given header <h> returns the number of bytes 
// allocated.
static inline unsigned ck_nbytes(hdr_t *h) {
    return h->nbytes_alloc;
}

// given header <h> return a pointer to the start of
// allocated data.
static inline void *ck_data_start(hdr_t *h) {
    return &h[1];
}

// given header <h> return pointer to the end of its
// allocated data.
static inline void *ck_data_end(hdr_t *h) {
    return (char *)ck_data_start(h) + ck_nbytes(h);
}

// pointer to first redzone [debug alloc lab]
static inline uint8_t *ck_get_rz1(hdr_t *h) {
    return h->rz1;
}

// pointer to the second redzon [debug alloc lab]
static inline uint8_t *ck_get_rz2(hdr_t *h) {
    return ck_data_end(h);
}

// is <ptr> in block <h>?
unsigned ck_ptr_in_block(hdr_t *h, void *ptr);

// returns header if pointer <p> on the allocated list?
hdr_t *ck_ptr_is_alloced(void *ptr);

// we do things this way so we can automatically pass in
// the location the routine was called at (using <SRC_LOC_MK()>)
// --- makes error reporting better.
#define ckalloc(_n) (ckalloc)(_n, SRC_LOC_MK())
#define ckfree(_ptr) (ckfree)(_ptr, SRC_LOC_MK())
void *(ckalloc)(uint32_t nbytes, src_loc_t loc);
void (ckfree)(void *addr, src_loc_t loc);

// integrity check the allocated / freed blocks in the heap
//
// returns number of errors in the heap.   stops checking
// if heap is in an unrecoverable state.
//
// TODO (extensions): 
//  - probably should have returned an error log, so that the 
//    caller could report / fix / etc.
//  - give an option to fix errors so that you can keep going.
int ck_heap_errors(void);

// public function: call to flag leaks (part 1 of lab).
//  - <warn_no_start_ref_p> = 1: warn about blocks that only have 
//    internal references.
//
//  - returns number of bytes leaked (where bytes = amount of bytes the
//    user explicitly allocated: does not include redzones, header, etc).
unsigned ck_find_leaks(int warn_no_start_ref_p);

// mark and sweep: works similarly to ck_find_leaks, manually
// frees any unreferenced blocks and resets state to FREED
// 
// Invariant:
//  - it should always be the case that after calling ck_gc(), 
//    immediately calling ck_find_leaks() should return 0 bytes 
//    found.
unsigned ck_gc(void);

// These two routines are just used for test cases.

// Expects no leaks.
void check_no_leak(void);
// Expects leaks.
unsigned check_should_leak(void);

#if 0
// info about the heap useful for checking.
struct heap_info {
    // original start of the heap.
    void *heap_start;
    // end of active heap (the next byte we would allocate)
    void *heap_end;

    // ckfree increments this on each free.
    unsigned nbytes_freed;
    // ckmalloc increments this on each free.
    unsigned nbytes_alloced;
};

struct heap_info heap_info(void);
#endif


// can use this to do debugging that you can turn off an on.
#define ck_debug(args...) \
 do { if(ck_verbose_p) debug("CK_DEBUG:" args); } while(0)

// just emits an error.
#define ck_error(_h, args...) do {      \
        trace("ERROR:");\
        printk(args);                    \
        hdr_print(_h);                  \
} while(0)

// emit error, then panic.
#define ck_panic(_h, args...) do {      \
        trace("ERROR:");\
        printk(args);                    \
        hdr_print(_h);                  \
        panic(args);                    \
} while(0)

// used to print out header values.
static void inline hdr_print(hdr_t *h) {
    trace("\tlogical block id=%u, [addr=%p] nbytes=%d\n", 
            h->block_id, 
            ck_data_start(h),
            h->nbytes_alloc);
    src_loc_t *l = &h->alloc_loc;
    if(l->file)
        trace("\tBlock allocated at: %s:%s:%d\n", l->file, l->func, l->lineno);

#if 0
    // next lab
    l = &h->free_loc;
    if(h->state == FREED && l->file)
        trace("\tBlock freed at: %s:%s:%d\n", l->file, l->func, l->lineno);
#endif
}

// control amount of printing.
extern unsigned ck_verbose_p;
static inline void ck_verbose_set(int v) {
    ck_verbose_p = v;
}

#endif
