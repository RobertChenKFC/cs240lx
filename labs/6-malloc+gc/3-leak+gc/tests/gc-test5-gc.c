// test that compaction works with doubly-linked list:
//  compact -- should find all bytes
//  compact again [should find 0]
//  check for leaks [should find 0]
#include "rpi.h"
#include "ckalloc.h"

// #define N 10
#define N 1

// allocate a linked list.  do a leak detection.
void *test(void) {
    struct list {
        int x; 
        struct list *next;
        struct list *prev;
    } *h = 0;

    for(int i = 0; i < N; i++) {
        struct list *e = ckalloc(sizeof *h);
        memset(e, 0, sizeof *e);
        e->x = 0;
        if(h) {
            assert(!h->prev);
            h->prev = e;
            e->next = h;
        }
        h = e;
    }
    // compiler optimizes this stuff away.
    // h= 0;
    // even this:
    //  memset(&h, 0, sizeof h);
    // perhaps if we put it in another file?
    trace("have a live pointer <h>: should not leak\n");
    check_no_leak();
    // return to try to get around compiler tricks.
    return h;
}

unsigned bar(void);

void notmain(void) {
    printk("GC test: checking that a doubly-linked list is handled.\n");

    // ck_verbose_set(0);
    // note: if we do a check after this, may miss b/c the value is laying around
    // on the stack or reg
    unsigned x = bar();

    // should detect leaks.
    trace("should leak\n");
    check_should_leak();

    // should compact everything: i don't know if this should be a trace --
    // the issue is that the compiler could leave some random values and it's
    // not under our control. 

    // some kind of corruption: compaction is trashing the output.
    // ck_verbose_set(1);

    trace("compacted = [nbytes=%d]\n", ck_gc());


    output("should find 0 bytes on second gc\n");
    unsigned n = ck_gc();
    if(n)
        panic("a second gc should free 0 bytes: found %d\n", n);

    trace("should find no leak\n");
    check_no_leak();
}

// mild attempt to make sure that the random values from <test> are 
// gone by the time we get back to main.
unsigned bar(void) {
#if 1
    void *p = test();
    put32(&p, 0);
    return 0;
#else
    void *p = ckalloc(4);
    ck_verbose_set(1);
    hdr_t *h = ck_ptr_is_alloced(p);
    if(h)
        output("about to check leak on block: %d, ptr=%x\n", h->block_id, p);

    check_should_leak();
    put32(p,0);
    return 0;

#endif
}
