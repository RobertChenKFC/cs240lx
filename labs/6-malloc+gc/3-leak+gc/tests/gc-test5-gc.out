TRACE:test:have a live pointer <h>: should not leak
TRACE:sweep_leak:	GC:Checked 1 blocks.
TRACE:sweep_leak:		GC:SUCCESS: No leaks found!
TRACE:check_no_leak:GC: SUCCESS: no leaks!
TRACE:notmain:should leak
TRACE:sweep_leak:ERROR:GC:DEFINITE LEAK of block=1 
TRACE:hdr_print:	logical block id=1,  nbytes=12
TRACE:hdr_print:	Block allocated at: tests/gc-test5-gc.c:test:20
TRACE:sweep_leak:	GC:Checked 1 blocks.
TRACE:sweep_leak:		GC:ERRORS: 1 errors, 0 maybe_errors
TRACE:check_should_leak:GC: SUCCESS: found 1 leaks!
TRACE:notmain:---------------------------------------------------------
TRACE:sweep_free:GC:FREEing block id=1 
TRACE:sweep_free:	GC:Checked 1 blocks, freed 1, 12 bytes
TRACE:sweep_free:	GC:Checked 0 blocks, freed 0, 0 bytes
TRACE:notmain:should find no leak
TRACE:sweep_leak:	GC:Checked 0 blocks.
TRACE:sweep_leak:		GC:SUCCESS: No leaks found!
TRACE:check_no_leak:GC: SUCCESS: no leaks!
